plutus-core/index-envs/src/Data/RandomAccessList/SkewBinary.hs

```diff
@@ -127,25 +131,47 @@ unsafeIndexOne (BHead w t ts) !i =
            then indexTree halfSize offset' t1
            else indexTree halfSize (offset' - halfSize) t2
 
--- 1-based
-safeIndexOne :: RAList a -> Word64 -> Maybe a
-safeIndexOne Nil _ = Nothing
-safeIndexOne (BHead w t ts) !i =
-    if i <= w
-    then indexTree w i t
-    else safeIndexOne ts (i-w)
-  where
-    indexTree :: Word64 -> Word64 -> Tree a -> Maybe a
-    indexTree _ 0 _ = Nothing -- "index zero"
-    indexTree 1 1 (Leaf x) = Just x
-    indexTree _ _ (Leaf _) = Nothing
-    indexTree _ 1 (Node x _ _) = Just x
+{- Note [Optimizations of safeIndexOneCont]
+Bangs in the local definitions of 'safeIndexOneCont' are needed to tell GHC that the functions are
+strict in the 'Word64' argument, so that GHC produces workers operating on @Word64#@.
+
+The function itself is CPS-ed, so that the arguments force the local definitions to be retained
+within 'safeIndexOneCont' instead of being pulled out via full-laziness or some other optimization
+pass. This ensures that when 'safeIndexOneCont' gets inlined, the local definitions appear directly
+in the GHC Core, allowing GHC to inline the arguments of 'safeIndexOneCont' and transform the whole
+thing into a beautiful recursive join point full of @Word64#@s, i.e. allocating very little if
+anything at all.
+-}
+
+-- See Note [Optimizations of safeIndexOneCont].
+safeIndexOneCont :: forall a b. b -> (a -> b) -> RAList a -> Word64 -> b
+safeIndexOneCont z f = findTree where
+    findTree :: RAList a -> Word64 -> b
+    -- See Note [Optimizations of safeIndexOneCont].
+    findTree Nil !_ = z
```

Comment:
Can you explain why you need to bang the wildcard here: `findTree Nil !_`

Reply:
It's in the referenced Note:

> Bangs in the local definitions of 'safeIndexOneCont' are needed to tell GHC that the functions are
> strict in the 'Word64' argument, so that GHC produces workers operating on @Word64#@.

Otherwise GHC doesn't know that it can unbox the second argument.